packages/@lwc/engine-core/src/framework/mutation-logger.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+//
+// Do additional mutation tracking for DevTools performance profiling, in dev mode only.
+//
+import {
+    ArrayPush,
+    ArraySplice,
+    isUndefined,
+    toString,
+    isObject,
+    isNull,
+    isArray,
+    keys,
+} from '@lwc/shared';
+import { ReactiveObserver } from '../libs/mutation-tracker';
+import { VM } from './vm';
+import { assertNotProd } from './utils';
+
+export interface MutationLog {
+    vm: VM;
+    prop: string;
+}
+
+const reactiveObserversToVMs = new WeakMap<ReactiveObserver, VM>();
+const targetsToPropertyKeys = new WeakMap<object, PropertyKey>();
+let mutationLogs: MutationLog[] = [];
+
+/**
+ * Flush all the logs we've written so far and return the current logs.
+ */
+export function getAndFlushMutationLogs() {
+    assertNotProd();
+    const result = mutationLogs;
+    mutationLogs = [];
+    return result;
+}
+
+/**
+ * Log a new mutation for this reactive observer.
+ * @param reactiveObserver - relevant ReactiveObserver
+ * @param target - target object that is being observed
+ * @param key - key (property) that was mutated
+ */
+export function logMutation(reactiveObserver: ReactiveObserver, target: object, key: PropertyKey) {
+    assertNotProd();
+    const parentKey = targetsToPropertyKeys.get(target);
+    const vm = reactiveObserversToVMs.get(reactiveObserver);
+
+    /* istanbul ignore if */
+    if (isUndefined(vm)) {
+        // VM should only be undefined in vitest tests, where a reactive observer is not always associated with a VM
+        // because the unit tests just create Reactive Observers on-the-fly.
+        if (process.env.NODE_ENV === 'test-karma-lwc') {
+            throw new Error('The VM should always be defined except possibly in unit tests');
+        }
+    } else {
+        // Human-readable prop like `items[0].name` on a deep object/array
```

Comment:
Actually we don't need to handle this, because (today at least) there's no way to reference `object["prop with spaces"]` in a template, so it won't be reactive.

We could revisit this after CTE though so that you don't see things like `object.prop with spaces` in DevTools.

Reply:
Ah wait, I'm wrong, you can do this with getters. I'll fix it.

Reply:
a1138bae1

Reply:
That's true! But we may as well make the label pretty. 🤓